Update for NAN@2 (incomplete)

rvagg opened the pull request:
The beginings of a NAN@2 upgrade, not complete cause it segfaults, someone else is willing to take over this if they have the patience to work out why. It compiles at least.

#173

rvagg commented:
I couldn't leave it alone, it works now but I've only tested on Node 4 so I have no idea if this plays nicely with older Node and I don't use contextify so ymmv.

ChALkeR commented:
Adding to the list: https://github.com/nodejs/node/issues/2798.

mzgol commented:
Related: https://github.com/brianmcd/contextify/issues/180#issuecomment-139055730

brianmcd commented:
Thank you so much, @rvagg! I'll do some regression testing on this later tonight and hopefully get a new version pushed out soon.

reggi commented:
Really happy progress is being made to fix this. Thanks everyone. 

I just tried to install @rvagg's version and it's not installing on node.js v4.0.0 here's my [`npm-debug.log`](https://gist.github.com/reggi/32da7944a1b5aae65a84).

brianmcd commented:
@rvagg - Thanks again for your help on this.  It works like a charm on 4.0.0, but I get the following errors on older versions:

- 0.12.7: https://gist.github.com/brianmcd/b99c9f4110ab9a3a5449
- 0.10.40: https://gist.github.com/brianmcd/388893aafd76c2b48e78

Any chance you'd be able to help out there?  I added travis to this repo with the tests running against node 4, 0.12, and 0.10, so it should make multi-version testing a little less painful.  Another push here should trigger it.

rvagg commented:
Sorry, I was using `v8::MaybeLocal` (implicit `v8` namespace) which doesn't exist on those versions, switched to `Nan::MaybeLocal`. I've fixed that now and it _mostly_ works with one remaining hiccup in the tests for v0.10.40. I've also added a .dntrc file just in case you're interested in testing with [dnt](https://github.com/rvagg/dnt/).

```
$ dnt
node@v0.10.40: FAIL wrote output to /tmp/contextify-dnt-v0.10.40.out
node@v0.12.7 : PASS wrote output to /tmp/contextify-dnt-v0.12.7.out
node@v1.8.4  : PASS wrote output to /tmp/contextify-dnt-v1.8.4.out
node@v2.5.0  : PASS wrote output to /tmp/contextify-dnt-v2.5.0.out
node@v3.3.0  : PASS wrote output to /tmp/contextify-dnt-v3.3.0.out
node@v4.0.0  : PASS wrote output to /tmp/contextify-dnt-v4.0.0.out
```

@kkoopa I'm at a loss on this one, only failing on v0.10.40. It's failing on these:

```js
        test.throws(function () {
            sandbox.run('doh');
        }, ReferenceError);
        test.throws(function () {
            sandbox.run('x = y');
        }, ReferenceError);
```

and when you inspect what it does, at line 98 in contextify.cc, this:

```c++
        Nan::MaybeLocal<Value> result = Nan::RunScript(script.ToLocalChecked());
```

is returning an `undefined` but is not empty and the `TryCatch` isn't catching anything. As far as I can tell it's doing exactly the same thing as it was with NAN@1 (via `Handle<Value> result = NanRunScript(script);`) but now it's not passing the error back up so I suppose it's to do with other infrastructure in place around it. Ideas?

kkoopa commented:
Don't know what's up. It should be hitting https://github.com/nodejs/nan/blob/master/nan.h#L784-L788

```c++
NAN_INLINE MaybeLocal<v8::Value> RunScript(
    v8::Local<BoundScript> script
) {
  return MaybeLocal<v8::Value>(script->Run());
}
```

sirbrillig commented:
FWIW, works for me in node 0.12.7. So I guess the issue is still just on 0.10.x?
